### ISSUE TITLE
Empty "todo list" calendars shown too presently

### ISSUE THREAD

**Issue by benyanke** [nextcloud/calendar#904]
It would be awesome to be able to hide calendars from view in the sidebar.

A significant portion of my calendars exist only to house named todo lists, and in my usecase, make no sense to show next to the calendar.

It would be nice to be able to move some of the calendars into a collapsed section which would effectively "hide" them.

**Comment by georgehrke** [comment]
> It would be awesome to be able to hide calendars from view in the sidebar.

Are you asking for a way to hide the sidebar altogether or just the calendars inside the sidebar?

**Comment by benyanke** [comment]
Hide specific items in the sidebar. I realize this is a bit of a hack with the way caldav works, but I end up with many calendars due to lists being shared.

For example: I have my personal todo on my main calendar, but I also have another calendar I created solely for the todo list called `ShoppingList` which I share with my wife. This - of course, makes no sense to show in the sidebar of the calendar app.

I know I can't be the only one using it this way.

**Comment by georgehrke** [comment]
@benyanke In that case I would vote against implementing this feature.
Calendars have certain component types. You can configure, whether they can store events, todos and journals. 

The current default (in the calendar app) is events and todos: https://github.com/nextcloud/calendar/blob/master/js/app/service/calendarService.js#L289

We should probably change this, so that calendars created in the calendar app can only store events, and calendars created in the tasks app can only store todos. (cc @jancborchardt @tcitworld @raimund-schluessler  @skjnldsv)

If the component type of a calendar does not contain events, it also won't show up in the left sidebar.

(If you want to change the component type of a calendar after creating it, you will have to edit the components column in the `oc_calendars` table of the calendar in question. Please be aware that changing the component type is not allowed per specification, so it might result in unexpected behavior in certain clients. There will be no unexpected behavior in the Nextcloud calendar / tasks app though)

**Comment by benyanke** [comment]
> We should probably change this, so that calendars created in the calendar app can only store events, and calendars created in the tasks app can only store todos. 

Perhaps it would be ideal to make it more configurable, and simply ask the user on calendar creation?

Something like:

"This calendar can store:
[ ] Events
[ ] Tasks
[ ] Notes
"

**Comment by tcitworld** [comment]
> We should probably change this, so that calendars created in the calendar app can only store events, and calendars created in the tasks app can only store todos

Hmm, so there wouldn't be any way to create a calendar which can also save todos from Nextcloud then ? And since the specification doesn't allow you to change the component allowed types, it also won't be changeable through 3rd party clients. So... I guess :-1: .

Allowing to choose at creation will result in friction for something 99% of users don't need to pick.

My opinion - based on my own usage of both Calendar & Tasks apps - would be that calendars created in the task app would only allow todos, and calendars created from the calendar app would allow both. The Calendar app wouldn't show calendars which support only todos.

**Comment by benyanke** [comment]
`My opinion - based on my own usage of both Calendar & Tasks apps - would be that calendars created in the task app would only allow todos, and calendars created from the calendar app would allow both. The Calendar app wouldn't show calendars which support only todos.`

Why not just allow the user to select which they want on calendar creation?

**Comment by tcitworld** [comment]
> Allowing to choose at creation will result in friction for something 99% of users don't need to pick.

**Comment by benyanke** [comment]
Not if it's hidden by default under 'more options' or something, with sane defaults. Especially since many users of nextcloud (more than other tools) lean harder down the power-user spectrum.

**Comment by jancborchardt** [comment]
How about just having a section for "Show all calendars" which includes calendars without any events in it?

We do this similarly in the mail app, where only the important folders are listed directly, and the others are behind a "Show all folders" toggle:
![nextcloud mail show all folders](https://user-images.githubusercontent.com/925062/49451589-cf26e300-f7df-11e8-8bc3-b747588f2dd3.png)

**Comment by tcitworld** [comment]
> calendars without any events in it?

It will be more like "without any events currently shown" (in the current event cache), since we don't want load all events.

**Comment by jancborchardt** [comment]
I would say we should do it like iCloud / macOS, simply having one list of calendars. Ones which are not your own or have no events could be hidden behind the "Show all calendars" toggle I mentioned we have in Mail.

**Comment by georgehrke** [comment]
@jancborchardt I think that would conflict with #9. If you have your own calendars and calendars shared with you mixed in one list, clicking "Show all calendars" might not only expand the list on the bottom but also add shared calendars in between your own calendars.

Also I'm not sure you can really compare Mail and Calendar in this case. With mail, it's obvious that "Inbox", "Favourites" and "Sent" are the most important. With calendar, there might be a group calendar that's used by everyone.

**Comment by jancborchardt** [comment]
@georgehrke ok, I’m also fine with what you proposed – you have the better overview here. :)

**Comment by georgehrke** [comment]
Talking about #9, one option could be to have one list, make it completely sortable, only show the top 10 and hide everything else behinds "Show all calendars"

**Comment by jancborchardt** [comment]
And default sorting simply by most recent activity. (Although if activity inside of the same day, their sorting should be alphabetic to not move around so much.) That way the most relevant calendars are simply always there.

**Comment by georgehrke** [comment]
Current solution:
- merge calendar and subscription list
- make it sortable #9 
- only show first 10 calendars and offer show more

**Comment by jancborchardt** [comment]
Somehow this seems to create more problems than there were before.
- Why limit the calendar list? It’s a scrolling list and it’s fine to have a bunch of entries in there.
- Why make it sortable? If it’s intelligently sorted, e.g. by most used (most events?) then the relevant stuff is always there.
- The calendars with no events would then be at the bottom automatically. Ta-da, problem solved. ;)

**Comment by tcitworld** [comment]
> Why make it sortable? If it’s intelligently sorted, e.g. by most used (most events?) then the relevant stuff is always there.
> The calendars with no events would then be at the bottom automatically. Ta-da, problem solved. ;)

Unfortunately we don't fetch all the events for a calendar, just those appropriate for the current view, therefore if we wanted to calculate which calendar has the most events for the current view it would be different at each view change.

I though `LAST-MODIFIED` on calendar objects from [RFC7986](https://tools.ietf.org/html/rfc7986#section-5.4) might have helped us but it only
> specifies the date and time that the information associated with the calendar was last revised

not the date of latest events added to it, which would be a poor criteria to sort calendars anyway.

Anyway, so there's currently no way to sort calendars intelligently.

**Comment by georgehrke** [comment]
I can only second what @tcitworld said. We don't (and we really really don't want to) fetch all events of a calendar.

macOS calendar / iCloud allows sorting.
Google puts your default calendar on top and sorts all the other calendars alphabetically.

**Comment by jancborchardt** [comment]
Adjusted the title of the issue to reflect the problem instead of a proposal. :)

**Comment by benyanke** [comment]
To add more info to this, if you create a calendar from within the TODO app, it creates a "calendar" which only stores VTODO objects, not VEVENT. This can be found in the `oc_calendars` table in the `components` field (`transparent` also seems to play into this?).

However, this is one case, and at create time.

I'd like to be able to, at any time, change the settings on a given calendar to implement VEVENT, VTODO, or both. Toggling on and off the capability. Toggling off would, I assume, lose the data, but you wouldn't turn it off unless you didn't want the data, so this would be doable with a warning.

**Comment by benyanke** [comment]
In the meantime, it appears you can adjust your existing ones using:

```sql
update components="VTODO",transparent="1" where id in(1,2,3);
```

Where 1,2, and 3 are IDs of calendars.

**Comment by benyanke** [comment]
Note: I believe this can already be accomplished with SQL, for those finding this in google and wanting a solution now:

```sql

; find your calendar IDs in  oc_calendars, put in the () list

; CHANGE TO TODO ONLY

update oc_calendars set components="VTODO",transparent="1" where id in (12,13,14);
delete from oc_calendarobjects where componenttype='VEVENT' and calendarid in (12,13,14);


; CHANGE TO EVENT ONLY


update oc_calendars set components="VEVENT",transparent="0" where id in (12,13,14);
delete from oc_calendarobjects where componenttype='VTODO' and calendarid in (12,13,14);

```

**Comment by raimund-schluessler** [comment]
> I'd like to be able to, at any time, change the settings on a given calendar to implement VEVENT, VTODO, or both. Toggling on and off the capability.

Changing the calendar components after the creation of a calendar is not allowed by the iCalendar specification, so this won't be implemented by the Calendar or Tasks apps. Please use the above code at your own risk.

**Comment by alexanderdd** [comment]
Isn't this fixed already? In calendar  2.3.1 in the settings at the bottom left I can toggle "show tasks" (German "Aufgaben im Kalender anzeigen"), which seems to accomplish what this issue is about.

**Comment by devvythelopper** [comment]
> Isn't this fixed already? In calendar 2.3.1 in the settings at the bottom left I can toggle "show tasks" (German "Aufgaben im Kalender anzeigen"), which seems to accomplish what this issue is about.

It would be nice if this was the default (tasks not shown in the calendar view) and you'd have to go to settings to show them anyways.
